btreeop/btreeop.1
.\"
.\" Copyright (c) 1996, 1997, 1998, 1999
.\"             Shigio Yamaguchi. All rights reserved.
.\" Copyright (c) 1999, 2000
.\"             Tama Communications Corporation. All rights reserved.
.\"
.\" This file is part of GNU GLOBAL.
.\"
.\" GNU GLOBAL is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" GNU GLOBAL is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.Dd Nov 3, 1998
.Dt BTREEOP 1
.Os FreeBSD
.Sh NAME
.Nm btreeop
.Nd btree database maintenance tool
.Sh SYNOPSIS
.Nm btreeop
.Op Fl A
.Op Fl C
.Op Fl D[keyno] Ar key
.Op Fl K[keyno] Ar key
.Op Fl L[2]
.Op Fl k Ar prefix
.Op Ar dbname
.Sh DESCRIPTION
.Nm Btreeop
execute simple operations for
.Xr btree 3
database.
.Nm Btreeop
can create database, write record, read record (sequential or indexed) and
delete record from it.
Duplicate entries are allowed.
Multi-key is available but only primary key can use index of
.Xr btree 3 .
.Sh OPTIONS
A capital letter means a command. If no command specified
then it assume sequential read operation.
.Bl -tag -width Ds
.It Fl A
Append records. If database doesn't exist, btreeop creates it.
.It Fl C
Create database and write records to it.
.It Fl D[keyno] Ar key
Delete records by the key. By default, keyno is 0 (primary key).
.It Fl K[keyno] Ar key
Search records by the key. By default, keyno is 0 (primary key).
.It Fl L[2]
List all primary keys. If '2' is specified, list all the key and data pairs.
.It Fl k Ar prefix
Scan records which have the prefix as a primary key.
This option is valid only with sequential read operation (-L command or non command).
.It Ar dbname
Database name. default is 'btree'.
.Sh DATA FORMAT
To creat (or append) database,
.Nm btreeop
read data from stdin.
The format of the data is the following.

  Primary-key  Secondary-key-1  Secondary-key-2  Data
  (keyno = 0)  (keyno = 1)      (keyno = 2)
  --------------------------------------------------------------
  main         246              main.c           main (){\\n
  func         120              library.c        func(a1, a2)\\n
  .
  .
  .

.El

.Bl -enum -offset indent
.It
All key and data are separated by blank('\\t' or ' '). 
.It
Key cannot include blank.
.It
Data can include blank.
.It
Null data not allowed.
.It
Additionally, META record is available. META record has a key that start with
a blank. You can read this record only by indexed read operation (-K option).
Usage is unlimited by btreeop.
.El
.Sh EXAMPLES
Create database.

  % btreeop -C
  key1	data1
  key2	data2
  key3	data3
  ^D
  %

Append records.

  % btreeop -A
   __.VERSION 2
  key2	data2-2
  ^D
  %

Sequential read.

  % btreeop
  key1  data1
  key2  data2-2
  key2  data2
  key3  data3
  %

Indexed read.

  % btreeop -K key2
  key2  data2-2
  key2  data2
  % btreeop -K ' __.VERSION'
   __.VERSION 2
  %

List primary keys.

  % btreeop -L
  key1
  key2
  key3
  %

Delete record.

  % btreeop -D ' __.VERSION'
  % btreeop -K ' __.VERSION'
  %

.Sh FILES
.Bl -tag -width tags -compact
.It Pa btree
default database name.
.El
.Sh DIAGNOSTICS
.Nm Btreeop
exits with a non 0 value if an error occurred, 0 otherwise.
.Sh SEE ALSO
.Xr btree 3
.Sh AUTHOR
Shigio Yamaguchi (shigio@tamacom.com)
.Sh HISTORY
The
.Nm
command appeared in FreeBSD 2.2.2.


NEWS
___________________________________
|      |  |  |     |  _  |     |  |
|  |___|  |  |  |  |    _|  |  |  |    GNU GLOBAL source code tagging system
|  |   |  |  |  |  |     |     |  |
|  ~~  |   ~~|     |  ~  |  |  |   ~~|          for all hackers.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Copyright (c) 2000-2016 Tama Communications Corporation

 This file is free software; as a special exception the author gives
 unlimited permission to copy and/or distribute it, with or without
 modifications, as long as this notice is preserved.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
 implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

		----------------------------------

GNU GLOBAL NEWS - User visible changes.
At least, you should search '[INCOMPATIBLE CHANGES]' for each version.

Version 6.5.7 - 15 May 2017

[CHANGES]
o global: Now global(1) accept a file as --nearness option.
  If it is a file, tags in the file are shown first.
o global: Added aliases for MAKEOBJDIR and MAKEOBJDIRPREFIX.
  GTAGSOBJDIR: alias for MAKEOBJDIR
  GTAGSOBJDIRPREFIX: alias for MAKEOBJDIRPREFIX
  These aliases are given higher priority than the originals.
o global: New command --print=<name>.
  <name> may be one of: root, dbpath or conf.
  This is a generic version of the -p (--print-dbpath) command.
	'global -p' is equal to 'global --print=dbpath'
	'global -pr' is equal to 'global --print=root'
o gtags: Updated the built-in PHP parser to support namespaces and traits,
  which were introduced in PHP5. The following keywords have been removed:
	- failure: not a keyword
	- header: normal built-in function
	- is_array: normal built-in function
	- is_set: typo -> isset
	- printf: normal built-in function
	- setcookie: normal built-in function
	- success: not a keyword

[FIXED BUGS]
o globash: Echo command with -n option didn't work on some OS. Now it works.
o global: The --nearness option without -x option brought segmentation fault.
	$ global --nearness=. main
	Segmentation fault: 11
  Now it works.
o global: --from-here with -i option did not ignore case of function name
  in the context. Now it works.
	Input:
	[main.c]
	1: main() {
	2:         func();
	3:         Func();
	4: }
	5: func() {}
	6: Func() {}

	$ global -xi --from-here=6:main.c func
	(global-6.5.6)
	Func                6 main.c           Func() {}
	func                5 main.c           func() {}
	(global-6.5.7)
	Func                3 main.c            Func();
	func                2 main.c            func();

[INCOMPATIBLE CHANGES]
o gtags.conf: Added new candidate path.
        (1) --gtagsconf file
        (2) $GTAGSCONF
        (3) [project root]/gtags.conf
        (4) [project root]/[objdir]/gtags.conf       <- [NEW]
        (5) $HOME/.globalrc
        (6) /etc/gtags.conf
        (7) [sysconfdir]/gtags.conf

        The default of [objdir] is 'obj'. You can change it by
        the environment variables (GTAGSOBJDIR or MAKEOBJDIR).

Version 6.5.6 - December 19 2016

[CHANGES]
New facilities:
o htags-server: New --retry[=n] option.
  If the port is already in use, retry n times with incrementing the port number.
  The default of n is 20.
o htags: Changed the format of function header (--show-position) to make copying
  text easier.
o geco.rc: Added 'fzf' as a selector candidate.
o gtags: New configuration variable 'gtags_hook'.
  gtags_hook=<command line>
	Specify a command line which should be executed at the beginning of gtags(1).
	Leading "./" in any path is always means the project root directory, since
	gtags(1) is always invoked there.
  GTAGS_COMMANDLINE environment variable (read only)
	You can get the effective command line of gtags(1) from the hook. It includes
	both $GTAGS_OPTIONS and real arguments.

   [Usage]
   You can update 'gtags.files' before gtags(1) read it.

	[gtags.conf]
	+----------------------------------------
	|...
	|:gtags_hook=find src lib -print >gtags.files:

   You can refer the effective arguments of gtags(1) from the hook using
   environment variable GTAGS_COMMANDLINE.

	[gtags.conf]
	+----------------------------------------
	|...
	|:gtags_hook=./gen.sh:\
	|:GTAGS_OPTIONS=-c:\

	[gen.sh]
	+----------------------------------------
	|#!/bin/sh
	|echo ">>> $GTAGS_COMMANDLINE"		# show effective command line

	$ gtags -O
	>>> gtags -c -O
	$ _

[INCOMPATIBLE CHANGES]
o htags: Now --cvsweb option always insert 'view=log' to the generated URLs.
  Because it seems to be almost always a desirable specification.

[FIXED BUGS]
o gtags: Gtags often aborts with a message "buffer overflow. strlimcpy" when
  it encounters a long token (> 152 bytes). Now gtags always ignores it with
  a message "symbol name is too long. (Ignored)".
o htags: Old packages included two CGI scripts (completion.cgi, global.cgi)
  generated in the release manager's machine by mistake. They have some literal
  path like '/opt/local/bin/perl' which works only with MacPorts.
  Now, they are generated in the target (your) machine.

Version 6.5.5 - Sep 21 2016

This is a bug fix release.

[FIXED BUGS]
o exuberant-ctags.la: The path name of ctags command couldn't changed.
  Now, it can be changed by configuration variable 'ctagscom'.
o Tag files got corrupted when the built-in C++ parser gave up parsing and aborted.
  Now, it is avoided.

o Some inappropriate error messages were changed.

  Error1: NULL file (file size == 0)

	$ cp /dev/null GTAGS
	$ global -x main

	(GLOBAL-6.5.4)	global: /tmp/test/GTAGS seems older format. Please remake tag files.
	(GLOBAL-6.5.5)	global: /tmp/test/GTAGS seems to be corrupted.

  Error2: Corrupted file

	$ echo aaa >GTAGS
	$ global -x main

	(GLOBAL-6.5.4)	global: GTAGS not found.
	(GLOBAL-6.5.5)	global: /tmp/test/GTAGS seems to be corrupted.

Version 6.5.4 - Mar 27 2016

This is a bug fix release.

[CHANGES]
Changed:
o gtags.conf: The entries which include '.xml' were commented out,

btreeop/btreeop.c
/*
 * Copyright (c) 1996, 1997, 1998, 1999
 *             Shigio Yamaguchi. All rights reserved.
 * Copyright (c) 1999, 2000
 *             Tama Communications Corporation. All rights reserved.
 *
 * This file is part of GNU GLOBAL.
 *
 * GNU GLOBAL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * GNU GLOBAL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <sys/types.h>
#include <sys/stat.h>

#include <ctype.h>
#include <signal.h>
#include <stdio.h>
#ifdef STDC_HEADERS
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#else
#include <strings.h>
#endif


#include "gparam.h"
#include "die.h"
#include "dbop.h"
#include "strbuf.h"
#include "tab.h"
#include "version.h"

const char *dbdefault = "btree";   	/* default database name */
const char *progname  = "btreeop";		/* command name */
int statistics;

static void	usage(void);
void	signal_setup(void);
void	onintr(int);
int	main(int, char **);
void	dbwrite(DBOP *);
void	dbkey(DBOP *, char *, int);
void	dbscan(DBOP *, char *, int);
void	dbdel(DBOP *, char *, int);
void	dbbysecondkey(DBOP *, int, char *, int);

#define F_KEY	0
#define F_DEL	1

static void
usage()
{
	fprintf(stderr, "%s\n",
		"usage: btreeop [-A][-C][-D[n] key][-K[n] key][-L[2]][-k prefix][dbname]");
	exit(2);
}

/*
 * Btreeop catch signal even if the parent ignore it.
 */
int	exitflag = 0;

void
onintr(signo)
int	signo;
{
	signo = 0;	/* to satisfy compiler */
	exitflag = 1;
}

void
signal_setup()
{
	signal(SIGINT, onintr);
	signal(SIGTERM, onintr);
#ifdef SIGHUP
	signal(SIGHUP, onintr);
#endif
#ifdef SIGQUIT
	signal(SIGQUIT, onintr);
#endif
}

int
main(argc, argv)
int	argc;
char	*argv[];
{
	char	command = 'R';
	char	*key = NULL;
	int     mode = 0;
	const char *db_name;
	DBOP	*dbop;
	int	i, c;
	int	secondkey = 0;
	int	keylist = 0;
	char	*prefix = NULL;

	for (i = 1; i < argc && argv[i][0] == '-'; ++i) {
		if (!strcmp(argv[i], "--version"))
			version(NULL, 0);
#ifdef STATISTICS
		else if (!strcmp(argv[i], "--dump")) {
			command = 'P';
			continue;
		}
		else if (!strcmp(argv[i], "--stat")) {
			statistics = 1;
			continue;
		}
#endif
		switch (c = argv[i][1]) {
		case 'D':
		case 'K':
			if (argv[i][2] && isdigit(argv[i][2]))
				secondkey = atoi(&argv[i][2]);
			if (++i < argc)
				key = argv[i];
			else
				usage();
			/* FALLTHROUGH */
		case 'A':
		case 'C':
		case 'L':
			if (command != 'R')
				usage();
			command = c;
			if (command == 'L') {
				keylist = 1;
				if (argv[i][2] == '2')
					keylist = 2;
			}
			break;
		case 'k':
			if (++i < argc)
				prefix = argv[i];
			else
				usage();
			break;
		default:
			usage();
			break;
		}
	}
	db_name = (i < argc) ? argv[i] : dbdefault;
	switch (command) {
	case 'A':
	case 'D':
		mode = 2;
		break;
	case 'C':
		mode = 1;
		break;
	case 'K':
	case 'L':
	case 'P':
	case 'R':
		mode = 0;
		break;
	}
	dbop = dbop_open(db_name, mode, 0644, DBOP_DUP);
	if (dbop == NULL) {
		switch (mode) {
		case 0:
		case 2:
			die("cannot open '%s'.", db_name);
			break;
		case 1:
			die("cannot create '%s'.", db_name);
			break;
		}
	}
	switch (command) {
	case 'A':			/* Append records */
	case 'C':			/* Create database */
		dbwrite(dbop);
		break;
	case 'D':			/* Delete records */
		dbdel(dbop, key, secondkey);
		break;
	case 'K':			/* Keyed (indexed) read */
		dbkey(dbop, key, secondkey);
		break;
	case 'R':			/* sequencial Read */
	case 'L':			/* primary key List */
		dbscan(dbop, prefix, keylist);
		break;
#ifdef STATISTICS
	case 'P':
		dbop_dump(dbop);
		break;
#endif
	}
#ifdef STATISTICS
	if (statistics)
		dbop_stat(dbop);
#endif
	dbop_close(dbop);

	return exitflag;
}
/*
 * dbwrite: write to database
 *
 *	i)	dbop		database
 */
void
dbwrite(dbop)
DBOP	*dbop;
{
	char	*p, *c;
	char	keybuf[MAXKEYLEN+1];
	STRBUF	*ib = strbuf_open(MAXBUFLEN);

	signal_setup();
	/*
	 * Input file format:
	 * +--------------------------------------------------
	 * |Primary-key	secondary-key-1 secondary-key-2 Data\n
	 * |Primary-key	secondary-key-1 secondary-key-2 Data\n
	 * 	.
	 * 	.
	 * - Keys and Data are separated by blank('\t' or ' '). 
	 * - Keys cannot include blank.
	 * - Data can include blank.
	 * - Null record not allowed.
	 * - Secondary-key is assumed as a part of data by db(3).
	 *
	 * META record:
	 * You can write meta record by making key start with a ' '.
	 * You can read this record only by indexed read ('-K' option).
	 * +------------------
	 * | __.VERSION 2
	 */
	while ((p = strbuf_fgets(ib, stdin, STRBUF_NOCRLF)) != NULL) {
		if (exitflag)
			break;
		c = p;
		if (*c == ' ') {			/* META record */
			if (*++c == ' ')
				die("key cannot include blanks.");
		}
		for (; *c && !isspace(*c); c++)		/* skip key part */
			;
		if (*c == 0)
			die("data part not found.");
		if (c - p > MAXKEYLEN)
			die("primary key too long.");
		strncpy(keybuf, p, c - p);		/* make key string */
		keybuf[c - p] = 0;
		for (; *c && isspace(*c); c++)		/* skip blanks */
			;
		if (*c == 0)
			die("data part is null.");
		dbop_put(dbop, keybuf, p, 0);
	}
	strbuf_close(ib);
}

/*
 * dbkey: Keyed search
 *
 *	i)	dbop		database
 *	i)	skey		key for search
 *	i)	secondkey	0: primary key, >0: secondary key
 */
void
dbkey(dbop, skey, secondkey)
DBOP	*dbop;
char	*skey;
int	secondkey;
{
	char	*p;

	if (!secondkey) {
		for (p = dbop_first(dbop, skey, NULL, 0); p; p = dbop_next(dbop))
			fprintf(stdout, "%s\n", p);
		return;
	}
	dbbysecondkey(dbop, F_KEY, skey, secondkey);
}

/*
 * dbscan: Scan records
 *
 *	i)	dbop		database
 *	i)	prefix		prefix of primary key
 *	i)	keylist		0: data, 1: key, 2: key and data
 */
void
dbscan(dbop, prefix, keylist)
DBOP	*dbop;
char	*prefix;
int	keylist;
{
	char	*p;
	int	flags = 0;

	if (prefix)	
		flags |= DBOP_PREFIX;
	if (keylist)
		flags |= DBOP_KEY;

	for (p = dbop_first(dbop, prefix, NULL, flags); p; p = dbop_next(dbop)) {
		if (keylist == 2)
			fprintf(stdout, "%s %s\n", p, dbop_lastdat(dbop));
		else
			detab(stdout, p);
	}
}

/*
 * dbdel: Delete records
 *
 *	i)	dbop		database
 *	i)	skey		key for search
 *	i)	secondkey	0: primary key, >0: secondary key
 */
void
dbdel(dbop, skey, secondkey)
DBOP	*dbop;
char	*skey;
int	secondkey;
{
	signal_setup();
	if (!secondkey) {
		dbop_delete(dbop, skey);
		return;
	}
	dbbysecondkey(dbop, F_DEL, skey, secondkey);
}
/*
 * dbbysecondkey: proc by second key
 *
 *	i)	dbop	database
 *	i)	func	F_KEY, F_DEL
 *	i)	skey
 *	i)	secondkey
 */
void
dbbysecondkey(dbop, func, skey, secondkey)
DBOP	*dbop;
int	func;
char	*skey;
int	secondkey;
{
	char	*c, *p;
	int	i;

	/* trim skey */
	for (c = skey; *c && isspace(*c); c++)
		;
	skey = c;
	for (c = skey+strlen(skey)-1; *c && isspace(*c); c--)
		*c = 0;

	for (p = dbop_first(dbop, NULL, NULL, 0); p; p = dbop_next(dbop)) {
		if (exitflag)
			break;
		c = p;
		/* reach to specified key */
		for (i = secondkey; i; i--) {
			for (; *c && !isspace(*c); c++)
				;
			if (*c == 0)
				die("specified key not found.");
			for (; *c && isspace(*c); c++)
				;
			if (*c == 0)
				die("specified key not found.");
		}
		i = strlen(skey);
		if (!strncmp(c, skey, i) && (*(c+i) == 0 || isspace(*(c+i)))) {
			switch (func) {
			case F_KEY:
				detab(stdout, p);
				break;
			case F_DEL:
				dbop_delete(dbop, NULL);
				break;
			}
		}
		if (exitflag)
			break;
	}
}
